src/sentry/static/sentry/app/views/organizationIntegrations/index.jsx

```diff
@@ -1,75 +1,141 @@
-import {Box} from 'grid-emotion';
+import {Box, Flex} from 'grid-emotion';
+import {keyBy} from 'lodash';
 import PropTypes from 'prop-types';
 import React from 'react';
 import styled from 'react-emotion';
 
+import {
+  Panel,
+  PanelBody,
+  PanelHeader,
+  PanelItem,
+  PanelItemGroup,
+} from 'app/components/panels';
+import {
+  addErrorMessage,
+  addLoadingMessage,
+  addSuccessMessage,
+} from 'app/actionCreators/indicator';
+import {openIntegrationDetails} from 'app/actionCreators/modal';
+import {sortArray} from 'app/utils';
 import {t} from 'app/locale';
-import {Panel, PanelBody, PanelHeader, PanelItem} from 'app/components/panels';
 import AsyncComponent from 'app/components/asyncComponent';
 import Button from 'app/components/buttons/button';
+import InstalledIntegration from 'app/views/organizationIntegrations/installedIntegration';
 import Link from 'app/components/link';
+import LoadingIndicator from 'app/components/loadingIndicator';
 import PluginIcon from 'app/plugins/components/pluginIcon';
 import SettingsPageHeader from 'app/views/settings/components/settingsPageHeader';
-import theme from 'app/utils/theme';
-
-const ProviderName = styled.div`
-  font-size: 16px;
-  font-weight: bold;
-  margin-bottom: 3px;
-`;
-
-const AuthorName = styled.div`
-  color: ${p => p.theme.gray2};
-  font-size: 14px;
-`;
 
 export default class OrganizationIntegrations extends AsyncComponent {
-  static propTypes = {
-    linkPrefix: PropTypes.string,
-    hideHeader: PropTypes.bool,
+  static contextTypes = {
+    organization: PropTypes.object.isRequired,
   };
 
+  // Some integrations require visiting a different website to add them. When
+  // we come back to the tab we want to show our integrations as soon as we can.
+  reloadOnVisible = true;
+
   getEndpoints() {
     let {orgId} = this.props.params;
     return [
       ['config', `/organizations/${orgId}/config/integrations/`],
-      ['organization', `/organizations/${orgId}/`],
+      ['integrations', `/organizations/${orgId}/integrations/`],
     ];
   }
 
+  mergeIntegration = integration => {
+    // Merge the new integration into the list. If we're updating an
+    // integration overwrite the old integration.
+    const keyedItems = keyBy(this.state.integrations, i => i.id);
+    const integrations = sortArray(
+      Object.values({...keyedItems, [integration.id]: integration}),
+      i => i.name
+    );
+    this.setState({integrations});
+  };
+
+  handleDeleteIntegration = integration => {
+    const {orgId} = this.props.params;
+    addLoadingMessage();
+
+    const options = {
+      method: 'DELETE',
+      success: () => {
+        const integrations = this.state.integrations.filter(i => i.id !== integration.id);
```

Comment:
imo we should optimistically update and remove the need for a loading message.

Reply:
Are we optimistic like this in other places?

Reply:
Yeah, although it's a mix right now, we should move towards optimistic

Reply:
Let's punt this for a follow up PR?

Reply:
It's cool, about to push up a commit for it.

Reply:
7a0780b